pkg/skaffold/schema/validation/validation.go

```diff
@@ -77,6 +78,81 @@ func validateImageNames(artifacts []*latest.Artifact) (errs []error) {
 	return
 }
 
+func validateArtifactDependencies(artifacts []*latest.Artifact) (errs []error) {
+	errs = append(errs, validateArtifactDependencyUniqueness(artifacts)...)
+	errs = append(errs, validateArtifactCircularDependencies(artifacts)...)
+	return
+}
+
+// validateArtifactDependencies makes sure all artifact dependencies are found and don't have cyclic references
+func validateArtifactCircularDependencies(artifacts []*latest.Artifact) (errs []error) {
+	m := make(map[string]*latest.Artifact)
+	for _, artifact := range artifacts {
+		m[artifact.ImageName] = artifact
+	}
+	visited := make(map[string]bool)
+	for _, artifact := range artifacts {
+		if err := dfs(artifact, visited, make(map[string]bool), m); err != nil {
+			errs = append(errs, err)
+			return
```

Comment:
since this is part of a larger error collection block that finishes all checks and surfaces all errors, should we not fail fast here and let this run its course?

Reply:
since this validation is for cycle detection, we'll get stuck on a specific cycle if it exists and it seemed non-trivial to exit that cycle and then try to find the other cycles. So we won't be able to collect all cycle violations at a go which is why I thought to just surface out the first violation and the user can fix it one by one. WDYT?

Reply:
@briandealwis WDYT?